### ISSUE TITLE
What should be the max epoch_number for training?

### ISSUE THREAD

**Issue by Rabia-Noureen** [Franck-Dernoncourt/NeuroNER#23]
What should be the max epoch_number value should be to train the model at the first run of main.py? I tried to change it by
```
python main.py --maximum_number_of_epochs=2 --token_pretrained_embedding_filepath=""
```
And after 3 or 4 runs the model starts training again. Should it stop at some point?

**Comment by Franck-Dernoncourt** [comment]
>And after 3 or 4 runs the model starts training again. Should it stop at some point?

With `python main.py --maximum_number_of_epochs=2 --token_pretrained_embedding_filepath=""` the model should train for two epochs only. I have just tried it on Ubuntu 14.04.4 LTS x64 and it works as expected. Did you make any change to the code, and do you use Windows 10?

**Comment by Rabia-Noureen** [comment]
Yes sir i am using windows 10 no i have not made any changes to the code (I just changed it to maximum_number_of_epochs = 2 in paramerters file) after epochs number 3 it again started from 0.

**Comment by Franck-Dernoncourt** [comment]
Interesting, could you please paste the output displayed in your terminal?

**Comment by Rabia-Noureen** [comment]
Okay let me run run that model again then i will paste the output.

**Comment by Rabia-Noureen** [comment]
Apologies from my side it looks like i was an older version of NeuroNER (which i got from some where else-Sorry about that) I have downloaded it again from github Now the epochs number stops at 2. :-) 

C:\Users\Zill-E-Huma\AppData\Local\Programs\Python\Python35\NeuroNER-master\src>python main.py --maximum_number_of_epochs=2 --token_pretrained_embedding_filepath=""
NeuroNER version: 1.0-dev
TensorFlow version: 1.1.0
NeuroNER version: 1.0-dev
TensorFlow version: 1.1.0
{'character_embedding_dimension': 25,
 'character_lstm_hidden_state_dimension': 25,
 'check_for_digits_replaced_with_zeros': 1,
 'check_for_lowercase': 1,
 'dataset_text_folder': '../data/conll2003/en',
 'debug': 0,
 'dropout_rate': 0.5,
 'experiment_name': 'test',
 'freeze_token_embeddings': 0,
 'gradient_clipping_value': 5.0,
 'learning_rate': 0.005,
 'load_only_pretrained_token_embeddings': 0,
 'main_evaluation_mode': 'conll',
 'maximum_number_of_epochs': 2,
 'number_of_cpu_threads': 8,
 'number_of_gpus': 0,
 'optimizer': 'sgd',
 'output_folder': '../output',
 'parameters_filepath': '.\\parameters.ini',
 'patience': 10,
 'plot_format': 'pdf',
 'pretrained_model_folder': '../trained_models/conll_2003_en',
 'reload_character_embeddings': 1,
 'reload_character_lstm': 1,
 'reload_crf': 1,
 'reload_feedforward': 1,
 'reload_token_embeddings': 1,
 'reload_token_lstm': 1,
 'remap_unknown_tokens_to_unk': 1,
 'spacylanguage': 'en',
 'tagging_format': 'bioes',
 'token_embedding_dimension': 100,
 'token_lstm_hidden_state_dimension': 100,
 'token_pretrained_embedding_filepath': '',
 'tokenizer': 'spacy',
 'train_model': 1,
 'use_character_lstm': 1,
 'use_crf': 1,
 'use_pretrained_model': 0,
 'verbose': 0}
Formatting train set from CONLL to BRAT... Done.
Converting CONLL from BIO to BIOES format... Done.
Formatting valid set from CONLL to BRAT... Done.
Converting CONLL from BIO to BIOES format... Done.
Formatting test set from CONLL to BRAT... Done.
Converting CONLL from BIO to BIOES format... Done.
Load dataset... done (15.12 seconds)

Starting epoch 0
Training completed in 0.00 seconds
Evaluate model on the train set
processed 203621 tokens with 23499 phrases; found: 175195 phrases; correct: 3996.
accuracy:   2.93%; precision:   2.28%; recall:  17.00%; FB1:   4.02
              LOC: precision:   2.21%; recall:  38.59%; FB1:   4.19  124437
             MISC: precision:   0.99%; recall:   0.67%; FB1:   0.80  2327
              ORG: precision:   5.03%; recall:  13.75%; FB1:   7.37  17263
              PER: precision:   1.12%; recall:   5.29%; FB1:   1.85  31168

Evaluate model on the valid set
processed 51362 tokens with 5942 phrases; found: 44549 phrases; correct: 1053.
accuracy:   2.86%; precision:   2.36%; recall:  17.72%; FB1:   4.17
              LOC: precision:   2.28%; recall:  40.66%; FB1:   4.32  32750
             MISC: precision:   0.62%; recall:   0.33%; FB1:   0.43  483
              ORG: precision:   5.73%; recall:  16.70%; FB1:   8.53  3908
              PER: precision:   1.07%; recall:   4.29%; FB1:   1.71  7408

Evaluate model on the test set
processed 46435 tokens with 5648 phrases; found: 39814 phrases; correct: 955.
accuracy:   2.88%; precision:   2.40%; recall:  16.91%; FB1:   4.20
              LOC: precision:   2.16%; recall:  36.09%; FB1:   4.08  27837
             MISC: precision:   1.10%; recall:   0.85%; FB1:   0.96  545
              ORG: precision:   6.48%; recall:  15.95%; FB1:   9.21  4092
              PER: precision:   1.12%; recall:   5.07%; FB1:   1.83  7340

Generating plots for the train set
Generating plots for the valid set
Generating plots for the test set
shell_command: perl .\conlleval < ../output\en_2017-06-19_00-39-42-593139\000_train.txt > ../output\en_2017-06-19_00-39-42-593139\000_train.txt_conll_evaluation.txt
shell_command: perl .\conlleval < ../output\en_2017-06-19_00-39-42-593139\000_valid.txt > ../output\en_2017-06-19_00-39-42-593139\000_valid.txt_conll_evaluation.txt
shell_command: perl .\conlleval < ../output\en_2017-06-19_00-39-42-593139\000_test.txt > ../output\en_2017-06-19_00-39-42-593139\000_test.txt_conll_evaluation.txt
Formatting 000_train set from CONLL to BRAT... Done.
Formatting 000_valid set from CONLL to BRAT... Done.
Formatting 000_test set from CONLL to BRAT... Done.
The last 0 epochs have not shown improvements on the validation set.

Starting epoch 1
Training completed in 752.09 seconds
Evaluate model on the train set
processed 203621 tokens with 23499 phrases; found: 23957 phrases; correct: 13395.
accuracy:  91.93%; precision:  55.91%; recall:  57.00%; FB1:  56.45
              LOC: precision:  69.08%; recall:  76.90%; FB1:  72.78  7949
             MISC: precision:  45.38%; recall:  29.17%; FB1:  35.52  2210
              ORG: precision:  44.09%; recall:  48.76%; FB1:  46.30  6991
              PER: precision:  56.10%; recall:  57.86%; FB1:  56.97  6807

Evaluate model on the valid set
processed 51362 tokens with 5942 phrases; found: 6110 phrases; correct: 3434.
accuracy:  91.86%; precision:  56.20%; recall:  57.79%; FB1:  56.99
              LOC: precision:  65.06%; recall:  77.35%; FB1:  70.68  2184
             MISC: precision:  50.28%; recall:  28.96%; FB1:  36.75  531
              ORG: precision:  38.94%; recall:  46.98%; FB1:  42.58  1618
              PER: precision:  62.80%; recall:  60.59%; FB1:  61.67  1777

Evaluate model on the test set
processed 46435 tokens with 5648 phrases; found: 5982 phrases; correct: 3225.
accuracy:  90.68%; precision:  53.91%; recall:  57.10%; FB1:  55.46
              LOC: precision:  59.66%; recall:  77.22%; FB1:  67.31  2159
             MISC: precision:  48.53%; recall:  23.50%; FB1:  31.67  340
              ORG: precision:  44.93%; recall:  48.83%; FB1:  46.80  1805
              PER: precision:  57.27%; recall:  59.43%; FB1:  58.33  1678

Generating plots for the train set
Generating plots for the valid set
Generating plots for the test set
shell_command: perl .\conlleval < ../output\en_2017-06-19_00-39-42-593139\001_train.txt > ../output\en_2017-06-19_00-39-42-593139\001_train.txt_conll_evaluation.txt
shell_command: perl .\conlleval < ../output\en_2017-06-19_00-39-42-593139\001_valid.txt > ../output\en_2017-06-19_00-39-42-593139\001_valid.txt_conll_evaluation.txt
shell_command: perl .\conlleval < ../output\en_2017-06-19_00-39-42-593139\001_test.txt > ../output\en_2017-06-19_00-39-42-593139\001_test.txt_conll_evaluation.txt
Formatting 001_train set from CONLL to BRAT... Done.
Formatting 001_valid set from CONLL to BRAT... Done.
Formatting 001_test set from CONLL to BRAT... Done.
The last 0 epochs have not shown improvements on the validation set.

Starting epoch 2
Training completed in 773.46 secondsdone
Evaluate model on the train set
processed 203621 tokens with 23499 phrases; found: 23940 phrases; correct: 17651.
accuracy:  95.42%; precision:  73.73%; recall:  75.11%; FB1:  74.42
              LOC: precision:  81.49%; recall:  92.52%; FB1:  86.65  8107
             MISC: precision:  56.07%; recall:  68.12%; FB1:  61.51  4177
              ORG: precision:  71.44%; recall:  52.48%; FB1:  60.51  4643
              PER: precision:  76.80%; recall:  81.61%; FB1:  79.13  7013

Evaluate model on the valid set
processed 51362 tokens with 5942 phrases; found: 6083 phrases; correct: 4354.
accuracy:  94.91%; precision:  71.58%; recall:  73.27%; FB1:  72.42
              LOC: precision:  80.59%; recall:  86.77%; FB1:  83.56  1978
             MISC: precision:  54.15%; recall:  69.41%; FB1:  60.84  1182
              ORG: precision:  66.02%; recall:  45.94%; FB1:  54.18  933
              PER: precision:  75.58%; recall:  81.65%; FB1:  78.50  1990

Evaluate model on the test set
processed 46435 tokens with 5648 phrases; found: 5910 phrases; correct: 3921.
accuracy:  93.38%; precision:  66.35%; recall:  69.42%; FB1:  67.85
              LOC: precision:  76.55%; recall:  84.53%; FB1:  80.34  1842
             MISC: precision:  39.89%; recall:  62.68%; FB1:  48.75  1103
              ORG: precision:  66.37%; recall:  49.79%; FB1:  56.90  1246
              PER: precision:  72.37%; recall:  76.93%; FB1:  74.58  1719

Generating plots for the train set
Generating plots for the valid set
Generating plots for the test set
shell_command: perl .\conlleval < ../output\en_2017-06-19_00-39-42-593139\002_train.txt > ../output\en_2017-06-19_00-39-42-593139\002_train.txt_conll_evaluation.txt
shell_command: perl .\conlleval < ../output\en_2017-06-19_00-39-42-593139\002_valid.txt > ../output\en_2017-06-19_00-39-42-593139\002_valid.txt_conll_evaluation.txt
shell_command: perl .\conlleval < ../output\en_2017-06-19_00-39-42-593139\002_test.txt > ../output\en_2017-06-19_00-39-42-593139\002_test.txt_conll_evaluation.txt
Formatting 002_train set from CONLL to BRAT... Done.
Formatting 002_valid set from CONLL to BRAT... Done.
Formatting 002_test set from CONLL to BRAT... Done.
The last 0 epochs have not shown improvements on the validation set.
Finishing the experiment

**Comment by Rabia-Noureen** [comment]
I wanted to ask that what will be the max epochs number when i train the model first time which out setting the epochs number for the best results?

**Comment by Franck-Dernoncourt** [comment]
great maybe it also solves https://github.com/Franck-Dernoncourt/NeuroNER/issues/25 :-)

> I wanted to ask that what will be the max epochs number when i train the model first time which out setting the epochs number for the best results?

If you don't specify it as a parameter, then the max epochs number is defined in https://github.com/Franck-Dernoncourt/NeuroNER/blob/master/src/parameters.ini

**Comment by Rabia-Noureen** [comment]
Okay i got it thanks for your time
